pkg/virt-handler/container-disk/mount.go

```diff
@@ -609,3 +617,70 @@ func (m *mounter) unmountKernelArtifacts(vmi *v1.VirtualMachineInstance) error {
 
 	return fmt.Errorf("kernel artifacts record wasn't found")
 }
+
+func (m *mounter) getContainerDiskPath(vmi *v1.VirtualMachineInstance, volume *v1.Volume, volumeIndex int) (*safepath.Path, error) {
+	sock, err := m.socketPathGetter(vmi, volumeIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := m.podIsolationDetector.DetectForSocket(vmi, sock)
+	if err != nil {
+		return nil, fmt.Errorf("failed to detect socket for containerDisk %v: %v", volume.Name, err)
+	}
+
+	mountPoint, err := isolation.ParentPathForRootMount(m.nodeIsolationResult, res)
+	if err != nil {
+		return nil, fmt.Errorf("failed to detect root mount point of containerDisk %v on the node: %v", volume.Name, err)
+	}
+
+	return containerdisk.GetImage(mountPoint, volume.ContainerDisk.Path)
+}
+
+func computeChecksum(imageFile *safepath.Path) (uint32, error) {
+	var checksum uint32
+
+	err := imageFile.ExecuteNoFollow(func(path string) (err error) {
+		b, err := os.ReadFile(path)
+		if err != nil {
+			return err
+		}
+		checksum = crc32.ChecksumIEEE(b)
+		return nil
+	})
+
+	return checksum, err
+}
+
+func (m *mounter) ComputeChecksum(vmi *v1.VirtualMachineInstance, volume *v1.Volume, volumeIndex int) (uint32, error) {
+	containerdiskPath, err := m.getContainerDiskPath(vmi, volume, volumeIndex)
+	if err != nil {
+		return 0, err
+	}
+
+	return computeChecksum(containerdiskPath)
+}
+
+func (m *mounter) VerifyChecksums(vmi *v1.VirtualMachineInstance) error {
+	for i, volume := range vmi.Spec.Volumes {
+		if volume.VolumeSource.ContainerDisk != nil {
+			checksum, err := m.ComputeChecksum(vmi, &volume, i)
+			if err != nil {
+				return fmt.Errorf("failed to compute checksum for volume %s: %s", volume.Name, err)
+			}
+
+			for i, volumeStatus := range vmi.Status.VolumeStatus {
+				if volumeStatus.ContainerDiskVolume != nil && vmi.Status.VolumeStatus[i].Name == volume.Name {
+					expectedChecksum := vmi.Status.VolumeStatus[i].ContainerDiskVolume.Checksum
+
+					if expectedChecksum != 0 && expectedChecksum != checksum {
+						return fmt.Errorf("checksum error for containerdisk %s", volume.Name)
+					}
+					// checksum ok or not present
```

Comment:
I am a little bit worried about the update path. E.g. if we have a VM running with current KubeVirt v1.1.0, and we trigger an update which includes this change, then with crio >= 1.28 the implementation will fallback to using tags (since it cannot get the digest from the ImageID). I assume the `expectedChecksum` will not be present in the status, thus it will accept any containerdisk with the given tag without verification. Is it the case, or am I missing smth? Are we sure that volume statuses will be timely updated prior to any possible VM migration?

Reply:
@vasiliy-ul You are right. We also need to calculate checksum for already running VMs.

Reply:
Once this is backported virt-handler will compute checksums for already running VMs as this is done in the virt-handler sync-loop.

Reply:
Sorry, I misunderstood what you wrote, yes you're right we cannot guarantee that VMs will be updated before a live migration on crio 1.28 gets processed.

I'm afraid there's not much we can do about that aside from letting the community know about this issue, virt-handler will compute the checksum as soon as the sync loop executes but the probability of a race condition is there indeed.

Reply:
@acardace, can't we just abort the migration if the vmi does not have checksum in the status? I.e. wouldn't it work if we simpliy returned error here instead?

Reply:
@vasiliy-ul after discussing privately with @xpivarc we came up with the solution of just returning an error in the virt-handler sync loop without aborting the migration, this will make sure that the migration actually starts only when the source virt-handler has computed and stored the required checksums, otherwise, the controller execute() will fail and the key will be re-enqueued.

Reply:
I took a quick look and that seems aligned with what I suggested:
> I.e. wouldn't it work if we simpliy returned error here instead?

Thanks! :+1:

Reply:
Thank you!